src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -183,15 +192,15 @@ public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions
 
                     JavaScriptEncoder? encoder = options.Encoder;
 
-                    if (_nameCache.Count < NameCacheSizeSoftLimit)
+                    if (_nameCacheForWriting.Count < NameCacheSizeSoftLimit)
                     {
-                        formatted = _namingPolicy == null
-                            ? JsonEncodedText.Encode(original, encoder)
-                            : FormatEnumValue(original, encoder);
+                        formatted = JsonEncodedText.Encode(
+                            _namingPolicy == null ? original : FormatUsingNamingPolicy(original, _namingPolicy),
```

Comment:
Nit: given that this is a repeating pattern, it might make sense to have the `Format` method accept a nullable `JsonNamingPolicy` (and perhaps name it something like "FormatJsonName")